Changes for `tctl sso test`, `tctl sso configure` commands [SAML]

Tener opened the pull request:
These are necessary changes to support `tctl sso test` and (to a much smaller degree) `tctl sso configure` command. 

RFD: `tctl sso configure` command: https://github.com/gravitational/teleport/pull/9845
RFD: `tctl sso test` command: https://github.com/gravitational/teleport/pull/9775

See: https://github.com/gravitational/teleport/issues/9270 for original issue, which covers larger scope: SAML, OIDC and GitHub auth connectors. This PR is only touching on SAML, but the implementations for OIDC and GitHub should be parallel to this one.

Actual commands for SAML are implemented in: https://github.com/gravitational/teleport.e/pull/425

Webapps PR: https://github.com/gravitational/webapps/pull/717

Tener committed:
Changes for tctl sso test, tctl sso configure commands

Tener committed:
Merge branch 'master' into tener/tctl-sso

Tener committed:
Make SSO logging aware of test flows.

Tener committed:
Proper method comments.

Tener committed:
Linter issues.

Tener committed:
require additional permissions for SSO test flow.

Tener committed:
Update submodule

Tener commented:
Dear reviewers, but especially @r0mant @smallinsky : please take a close look at the security angle, including the topics mentioned in the respective RFDs. Thanks!

Tener committed:
Merge branch 'master' into tener/tctl-sso

Tener committed:
Correct the comment.

Tener committed:
Take into account possibly nil variable.

smallinsky commented on lib/auth/saml.go:
```diff
@@ -352,36 +412,62 @@ type samlAuthResponse struct {
 	if err != nil {
-		return nil, trace.Wrap(err)
+		traceErr("Failed to get SAML Auth Request", err)
+		return nil, request, trace.Wrap(err)
```
Generally it is bad practice to return a a valid value  from function that returns err != nil. 

Could you consider decoupling `request, err := a.Identity.GetSAMLAuthRequest(requestID)` from `validateSAMLResponse` flow  for example: 
```go
id, err := getReuqestIDFromSAMLResponse(samlResp)
request, err := a.Idenity.GetSAMLAuthRequest(id)
re, err := validateSAMLResponse(samlResponse, request) 
```

Tener replied:
In principle I agree, except we are already doing it here:

https://github.com/gravitational/teleport/blob/master/lib/auth/saml.go#L407

```go
	// Calculate (figure out name, roles, traits, session TTL) of user and
	// create the user in the backend.
	params, err := a.calculateSAMLUser(connector, *assertionInfo, request)
	if err != nil {
		return re, trace.Wrap(err)
	}
	user, err := a.createSAMLUser(params)
	if err != nil {
		return re, trace.Wrap(err)
	}
```

Any refactoring would have to change that too, and I'm not sure this is worth it.

Still, I'll give it some thought and experiment if there is a way to express this cleanly.

Tener replied:
I have refactored 'validateSAMLResponse' to make it only return an error or a value, but never both.

The cost is an additional parameter to the function. Personally, I find the new code easier to read, but let me know what you think.

smallinsky commented on lib/auth/saml.go:
```diff
@@ -128,6 +124,44 @@ func (a *Server) CreateSAMLAuthRequest(req services.SAMLAuthRequest) (*services.
+	var provider *saml2.SAMLServiceProvider
+	var err error
+
+	if req.SSOTestFlow {
```
nit: just a preference but for me the second form is more readable
```go
func name1() (int, int, err) {
	int x
	int z 
	int err
	if ... {
		x,err  = ...
		y,err = ...
	} else {
		x,err  = ...
		y,err = ...
	}
	return x, y , z
}
```
vs
```go
func name2() int, int, err {
	if ... {
		x,err  := ...
		y,err := ...
		return x, y, err
	} 
	x,err  := ...
	y,err := ...
	return x, y , z
}
```

Tener replied:
I'm not entirely happy with this function either, I'll give it a go.

Tener replied:
Done.

smallinsky commented on lib/auth/saml.go:
```diff
@@ -326,8 +372,14 @@ func (a *Server) ValidateSAMLResponse(samlResponse string) (*SAMLAuthResponse, e
+
 	if err != nil {
-		event.Code = events.UserSSOLoginFailureCode
+		if testFlow {
+			event.Code = events.UserSSOTestFlowLoginFailureCode
+		} else {
+			event.Code = events.UserSSOLoginFailureCode
+		}
```
https://github.com/uber-go/guide/blob/master/style.md#unnecessary-else

Tener replied:
Done.

smallinsky commented on lib/auth/saml.go:
```diff
@@ -336,14 +388,22 @@ func (a *Server) ValidateSAMLResponse(samlResponse string) (*SAMLAuthResponse, e
 	event.Status.Success = true
 	event.User = re.auth.Username
-	event.Code = events.UserSSOLoginCode
+	if testFlow {
+		event.Code = events.UserSSOTestFlowLoginCode
+	} else {
+		event.Code = events.UserSSOLoginCode
+	}
```
https://github.com/uber-go/guide/blob/master/style.md#unnecessary-else

Tener replied:
TBH I think this rule makes it **harder** to read code, not easier, but yeah, not gonna die on that hill.

smallinsky commented on lib/web/saml.go:
```diff
@@ -97,8 +99,22 @@ func (h *Handler) samlACS(w http.ResponseWriter, r *http.Request, p httprouter.P
 	if err != nil {
 		logger.WithError(err).Error("Error while processing callback.")
+
+		// try to find the auth request, which bears the original client redirect URL.
+		// if found, use it to terminate the flow.
+		//
+		// this improves the UX by terminating the failed SSO flow immediately, rather than hoping for a timeout.
```
Hm, I'm not sure why we the valid redirect URL is returned here instead of old `LoginFailedBadCallbackRedirectURL` url and where the actually timeout occurs. 
If the `LoginFailedBadCallbackRedirectURL` is returned the client should be able call local url instantly and exit failed SSO flow but if the request url.String() can be malformed and not reachable bu the client.  If some additional information needs to be passed to the client why not just  encode them in the same way  in `LoginFailedBadCallbackRedirectURL` ?

Tener replied:
If the SSO login is initiated from the console (e.g. `tsh login`) then ending the flow on `/web/msg/error/login/callback` instead of `ClientRedirectURL` causes the console to leave it hanging.

Eventually 3 minutes pass and [this happens](https://github.com/gravitational/teleport/blob/master/lib/client/weblogin.go#L319):

```go
	case <-time.After(defaults.CallbackTimeout):
		log.Debugf("Timed out waiting for callback after %v.", defaults.CallbackTimeout)
		return nil, trace.Wrap(trace.Errorf("timed out waiting for callback"))
```

The browser will get redirected to error URL by `Redirector` in `wrapCallback`: https://github.com/gravitational/teleport/blob/master/lib/client/redirect.go#L231

I'm not sure what is the exact change that you are proposing here?

Tener replied:
Good catch! Fixed: https://github.com/gravitational/teleport/pull/11508/commits/2ed6ad434a8263c53dff3bf60eae5c9ed10fd7b3

smallinsky commented on lib/client/redirect.go:
```diff
@@ -145,32 +150,48 @@ func (rd *Redirector) Start() error {
+	response, err := rd.IssueSSOLoginConsoleRequest(req)
 	if err != nil {
-		return trace.Wrap(err)
+		return err
```
```suggestion
		return trace.Wrap(err)
```

Tener replied:
Done.

smallinsky commented on lib/auth/github.go:
```diff
@@ -334,6 +334,28 @@ type createUserParams struct {
 	sessionTTL time.Duration
 }
 
+func (params createUserParams) toMap() map[string]interface{} {
```
Not sure why the params needs to be converted to map to be forwarded to `TraceSAMLDiagnosticInfo` as `extraInfo` param where `extraInfo` has the `interface{}` type.

Tener replied:
This is a workaround really as `createUserParams` has no exported fields and I don't want to change that type as it is used very widely.

Tener replied:
The function `toMap()` is no more, instead `createUserParams` has custom JSON marshaller now.

smallinsky reviewed (commented):
First pass, Left  some comments.

Tener commented on lib/auth/apiserver.go:
```diff
@@ -212,6 +212,8 @@ func NewAPIServer(config *APIConfig) (http.Handler, error) {
 	srv.POST("/:version/saml/requests/create", srv.withAuth(srv.createSAMLAuthRequest))
 	srv.POST("/:version/saml/requests/validate", srv.withAuth(srv.validateSAMLResponse))
+	srv.GET("/:version/saml/requests/get/:id", srv.withAuth(srv.getSAMLAuthRequest))
+	srv.GET("/:version/saml/requests/diag/:id", srv.withAuth(srv.getSAMLDiagnosticInfo))
```
RBAC is per SSO type so it does needs to be present somewhere, but it can be a parameter instead of route.

Tener replied:
Done.

Tener commented on lib/services/identity.go:
```diff
@@ -202,9 +202,15 @@ type Identity interface {
 	GetSAMLAuthRequest(id string) (*SAMLAuthRequest, error)
 
+	// TraceSAMLDiagnosticInfo creates new SAML diagnostic info record.
+	TraceSAMLDiagnosticInfo(ctx context.Context, authRequestID string, key string, value interface{}, extraInfo ...interface{}) error
```
I will change the signature and name.

Tener replied:
`Trace` => `Create`, signature updated.

Tener commented on lib/auth/saml.go:
```diff
@@ -352,36 +412,62 @@ type samlAuthResponse struct {
 
-func (a *Server) validateSAMLResponse(samlResponse string) (*samlAuthResponse, error) {
+const (
+	DiagInfoError                         = "common.error"
+	DiagInfoResult                        = "common.result"
+	DiagInfoCreateUserParams              = "common.createUserParams"
```
I'll find a better place to place these constants in.

Tener replied:
This is now a protobuf enum.

Tener commented on lib/events/codes.go:
```diff
@@ -64,6 +64,10 @@ const (
 	// recovery code was not used successfully.
 	RecoveryCodeUseFailureCode = "T1009W"
+	// UserSSOTestFlowLoginCode is the successful SSO test flow user login event code.
+	UserSSOTestFlowLoginCode = "T1010I"
+	// UserSSOTestFlowLoginFailureCode is the unsuccessful SSO test flow user login event code.
+	UserSSOTestFlowLoginFailureCode = "T1011W"
```
Huh, good to know. Yet another PR it seems.

Tener replied:
https://github.com/gravitational/webapps/pull/717

Tener commented on lib/auth/saml.go:
```diff
@@ -460,10 +568,12 @@ func (a *Server) validateSAMLResponse(samlResponse string) (*samlAuthResponse, e
 	}
 
-	return re, nil
+	_ = a.Identity.TraceSAMLDiagnosticInfo(ctx, requestID, DiagInfoResult, "success")
```
Yeah, this is true. They can be useful outside of SSO test flow too, but right now there is no other way of accessing them. 

@r0mant what do you think about giving admins a Web UI for diagnosing recent SSO login failures?

Tener committed:
Declare SSOInfoType as protobuf message.

Tener committed:
Make SSO diag info shared for all auth types.

Tener committed:
Rename function for clarity.

Tener committed:
Error handling fixes.

Tener committed:
Use custom marshalling for createUserParams type.

Tener committed:
Refactoring to match updated API, cleanups.

Tener commented on api/types/sso.go:
```diff
@@ -0,0 +1,24 @@
+package types
+
+// SsoDiagInfoEntry records diagnostic information about sso auth requests.
+type SsoDiagInfoEntry struct {
```
Done.

Tener commented on lib/auth/apiserver.go:
```diff
@@ -212,6 +212,8 @@ func NewAPIServer(config *APIConfig) (http.Handler, error) {
 	srv.POST("/:version/saml/requests/create", srv.withAuth(srv.createSAMLAuthRequest))
 	srv.POST("/:version/saml/requests/validate", srv.withAuth(srv.validateSAMLResponse))
+	srv.GET("/:version/saml/requests/get/:id", srv.withAuth(srv.getSAMLAuthRequest))
+	srv.GET("/:version/saml/requests/diag/:id", srv.withAuth(srv.getSAMLDiagnosticInfo))
```
SSO diagnostic info entries are now decoupled from SSO type.

Tener commented on lib/web/apiserver.go:
```diff
@@ -1331,6 +1331,19 @@ func ConstructSSHResponse(response AuthParams) (*url.URL, error) {
 	return u, nil
 }
 
+func returnErrorToClient(clientRedirectURL string, errReply error) (*url.URL, error) {
```
Renamed.

rosstimothy commented on api/types/sso_test.go:
```diff
@@ -0,0 +1,127 @@
+		Error   string
+	}
+
+	tests := []struct {
```
add a test case that results in an error?

Tener replied:
Done.

Tener commented on lib/auth/saml.go:
```diff
@@ -158,11 +192,17 @@ func (a *Server) calculateSAMLUser(connector types.SAMLConnector, assertionInfo
 	p.traits = services.SAMLAssertionsToTraits(assertionInfo)
 
+	_ = a.Identity.TraceSAMLDiagnosticInfo(context.Background(), request.ID, DiagInfoSAMLTraitsFromAssertions, p.traits)
+	_ = a.Identity.TraceSAMLDiagnosticInfo(context.Background(), request.ID, DiagInfoSAMLConnectorTraitMapping, connector.GetTraitMappings())
```
This is unchanged, but I see no strong reason to change the design.

rosstimothy commented on lib/auth/clt.go:
```diff
@@ -1753,6 +1779,12 @@ type IdentityService interface {
 	ValidateSAMLResponse(re string) (*SAMLAuthResponse, error)
 
+	// GetSAMLAuthRequest returns SAML auth request if found
+	GetSAMLAuthRequest(id string) (*services.SAMLAuthRequest, error)
```
The context should really be propagated here by the caller. This doesn't look like it is used frequently enough that adding this here and to the `IdentityService` will cause this PR to get too noisy.
```suggestion
	GetSAMLAuthRequest(ctx context.Context, id string) (*services.SAMLAuthRequest, error)
```

Tener replied:
Done for this and all methods calling `GetSAMLAuthRequest`.

rosstimothy commented on lib/auth/github.go:
```diff
@@ -334,6 +334,32 @@ type createUserParams struct {
+		SessionTTL:    cup.sessionTTL,
+	})
+	if err != nil {
+		return nil, err
```
```suggestion
		return nil, trace.Wrap(err)
```

Tener replied:
Done.

rosstimothy commented on lib/auth/saml.go:
```diff
@@ -128,6 +123,41 @@ func (a *Server) CreateSAMLAuthRequest(req services.SAMLAuthRequest) (*services.
+		// stateless test flow
+		connector, err := types.NewSAMLConnector(req.ConnectorID, *req.ConnectorSpec)
+		if err != nil {
+			return nil, nil, err
```
```suggestion
			return nil, nil, trace.Wrap(err)
```

Tener replied:
Done.

rosstimothy commented on lib/auth/saml.go:
```diff
@@ -128,6 +123,41 @@ func (a *Server) CreateSAMLAuthRequest(req services.SAMLAuthRequest) (*services.
+		// validate, set defaults for connector
+		err = services.ValidateSAMLConnector(connector)
+		if err != nil {
+			return nil, nil, err
```
```suggestion
			return nil, nil, trace.Wrap(err)
```

Tener replied:
Done.

rosstimothy commented on lib/auth/saml.go:
```diff
@@ -158,11 +188,36 @@ func (a *Server) calculateSAMLUser(connector types.SAMLConnector, assertionInfo
+			log.WithError(err).Warn("Failed to serialize SSO diag info.")
+		}
+
+		err = a.Identity.CreateSSODiagnosticInfo(context.Background(), types.KindSAML, request.ID, *entry)
```
The context here should either be a TODO, or the function should explicitly take a `context.Context` as the first parameter
```suggestion
		err = a.Identity.CreateSSODiagnosticInfo(context.TODO(), types.KindSAML, request.ID, *entry)
```

Tener replied:
I've modified the calls so that all callers are passing existing contexts.

rosstimothy commented on lib/auth/saml.go:
```diff
@@ -315,9 +374,13 @@ func (a *Server) ValidateSAMLResponse(samlResponse string) (*SAMLAuthResponse, e
-	if re != nil && re.attributeStatements != nil {
-		attributes, err := apievents.EncodeMapStrings(re.attributeStatements)
+
+	auxInfo := validateSAMLAuxInfo{}
```
```suggestion
	var auxInfo validateSAMLAuxInfo
```

Tener replied:
Done.

Tener committed:
Lint.

rosstimothy commented on lib/auth/saml.go:
```diff
@@ -336,87 +402,141 @@ func (a *Server) ValidateSAMLResponse(samlResponse string) (*SAMLAuthResponse, e
 
-	return &re.auth, nil
+	if err != nil {
+		return nil, err
```
```suggestion
		return nil, trace.Wrap(err)
```

Tener replied:
Done.

Tener committed:
Merge branch 'master' into tener/tctl-sso

rosstimothy commented on lib/client/redirect.go:
```diff
@@ -202,6 +223,11 @@ func (rd *Redirector) callback(w http.ResponseWriter, r *http.Request) (*auth.SS
 
+	if r.URL.Query().Has("err") {
+		err := r.URL.Query().Get("err")
+		return nil, trace.Errorf("received error: %v", err)
```
nit: `trace.Errorf` isn't generally used directly
```suggestion
		return nil, trace.Wrap(err)
```

Tener replied:
No can do: `err` is string here.

r0mant replied:
Is this a user facing error? If so, can the error message be made a bit clearer? E.g. "identity provider callback failed with the following error" or "error while handling the identity provider callback", or something.

rosstimothy commented on lib/services/saml.go:
```diff
@@ -124,33 +128,45 @@ func SAMLAssertionsToTraits(assertions saml2.AssertionInfo) map[string][]string
+func GetSAMLServiceProvider(sc types.SAMLConnector, clock clockwork.Clock) (*saml2.SAMLServiceProvider, error) {
+	roots, errEd := CheckSAMLEntityDescriptor(sc.GetEntityDescriptor())
+	if errEd != nil {
+		return nil, errEd
```
```suggestion
		return nil, trace.Wrap(errEd)
```

Tener replied:
Done.

Tener committed:
Add negative test case.

Tener committed:
Add missing argument - post-merge cleanup.

Tener committed:
Extend several methods with ctx param.

Tener committed:
Add missing trace.Wrap() calls

Tener committed:
Use existing ctx for CreateSSODiagnosticInfo call

Tener committed:
Use var.

Tener commented on lib/auth/saml.go:
```diff
@@ -336,87 +402,141 @@ func (a *Server) ValidateSAMLResponse(samlResponse string) (*SAMLAuthResponse, e
 	if err != nil {
-		return nil, trace.Wrap(err)
+		traceErr("Failed to get SAML connector and provider", err)
+		return nil, err
```
Done.

Tener committed:
Linter.

Tener committed:
Add missing license.

rosstimothy commented on api/types/sso.go:
```diff
@@ -0,0 +1,48 @@
+}
+
+// GetValue deserializes embedded JSON of SSODiagnosticInfo.Value with no assumption about underlying type.
+func (m *SSODiagnosticInfo) GetValue() (interface{}, error) {
+	var value interface{}
+	err := json.Unmarshal(m.Value, &value)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+	return value, nil
+}
+
+// GetValueTyped deserializes embedded JSON of SSODiagnosticInfo.Value given typed pointer.
+func (m *SSODiagnosticInfo) GetValueTyped(value interface{}) error {
+	return trace.Wrap(json.Unmarshal(m.Value, &value))
+}
```
Should `GetValue()` just be `GetValue(value interface{})` and `GetValueTyped` be removed? In which circumstances would the caller of this use `GetValue` over `GetValueTyped`?

rosstimothy replied:
I still think in that scenario it would make more sense to have `GetValue(value interface{})`, remove `GetValueTyped` and have the user explicitly pass in an `interface{}`

Tener replied:
> I still think in that scenario it would make more sense to have `GetValue(value interface{})`, remove `GetValueTyped` and have the user explicitly pass in an `interface{}`

Not a big deal, done.

Tener committed:
Replace GetValue with GetValueTyped.

smallinsky commented on api/types/types.proto:
```diff
@@ -2609,6 +2609,39 @@ message GithubConnectorSpecV3 {
+    // Create user params, contains info about attempted user creation parameters.
+    CREATE_USER_PARAMS = 3;
+    // [SAML] Attributes to roles mapping.
+    SAML_ATTRIBUTES_TO_ROLES = 4;
```
Do we want add some descriptor number space between common info types and known connector types ? so in case where a new type needed by added the values will be still aggregated.

Tener replied:
Sure, why not. Done: https://github.com/gravitational/teleport/pull/11508/commits/19c4c326dfb73dee2cf4adef94af1b11728c612f

Tener commented on lib/auth/saml.go:
```diff
@@ -158,11 +188,36 @@ func (a *Server) calculateSAMLUser(connector types.SAMLConnector, assertionInfo
+		}
+	}
+
+	diagInfo(types.SSOInfoType_SAML_TRAITS_FROM_ASSERTIONS, p.traits)
+	diagInfo(types.SSOInfoType_SAML_CONNECTOR_TRAIT_MAPPING, connector.GetTraitMappings())
```
See my other comment: https://github.com/gravitational/teleport/pull/11508#discussion_r841623570

We may want to follow up with admin UI to see recent logins.

r0mant replied:
Each login already generates an audit event. I think that we can safely skip generating diagnostic info for regular logins.

Tener replied:
I'll restrict the logging to test flow only. If we ever want to make the admin UI it would be simple to drop it.

Tener replied:
Done: e478da1c5ac88a1dbe5f21b4af20772dc980da82

Tener commented on api/types/types.proto:
```diff
@@ -2609,6 +2609,39 @@ message GithubConnectorSpecV3 {
+    SUCCESS = 2;
+    // Create user params, contains info about attempted user creation parameters.
+    CREATE_USER_PARAMS = 3;
+    // [SAML] Attributes to roles mapping.
```
Is this an actual requirement documented anywhere? What would be the rationale?

We don't follow it everywhere in .proto, e.g.: 
https://pkg.go.dev/github.com/gravitational/teleport/api/types#SessionState_SessionStatePending

```proto
// SessionState represents the state of a session.
enum SessionState {
    // Pending variant represents a session that is waiting on participants to fulfill the criteria
    // to start the session.
    SessionStatePending = 0;

    // Running variant represents a session that has had it's criteria for starting
    // fulfilled at least once and has transitioned to a RUNNING state.
    SessionStateRunning = 1;

    // Terminated variant represents a session that is no longer running and due for removal.
    SessionStateTerminated = 2;
}
```

Generated code looks like this:

```go
const (
	// Pending variant represents a session that is waiting on participants to fulfill the criteria
	// to start the session.
	SessionState_SessionStatePending SessionState = 0
	// Running variant represents a session that has had it's criteria for starting
	// fulfilled at least once and has transitioned to a RUNNING state.
	SessionState_SessionStateRunning SessionState = 1
	// Terminated variant represents a session that is no longer running and due for removal.
	SessionState_SessionStateTerminated SessionState = 2
)
```

So to truly follow godoc format we would need to write silly stuff like: 
```proto
    // SessionState_SessionStatePending Pending variant represents a session that is waiting on participants to fulfill the criteria
    // to start the session.
    SessionStatePending = 0;
```

smallinsky replied:
The type name generated by protobuf depends on if a type is global or embedded in another message so it is hard to keep this in sync  also the name has an unnecessary type prefix like `SessionState_SessionState`. 

I don't think that we have any document describing how proto fields should be documented but most of our proto docs comments borrow godocs style thus I think that is reasonable to keep it consistent but leaving it up to you.

Tener committed:
SSOInfoType: space out constants per type.

Tener committed:
Share SSO DiagInfo logic.

Tener committed:
Fix error redirect for web login

smallinsky commented:
@reedloden @russjones
Could you also take a look at this PR  from security perspective ?

r0mant commented on api/types/sso.go:
```diff
@@ -0,0 +1,38 @@
+)
+
+// NewSSODiagnosticInfo creates new SSODiagnosticInfo object using arbitrary value, which is serialized using JSON.
+func NewSSODiagnosticInfo(infoType SSOInfoType, value interface{}) (*SSODiagnosticInfo, error) {
```
Any reason we can't make this a properly typed object? In general we try to avoid using `interface{}` as much as possible.

r0mant commented on api/types/types.proto:
```diff
@@ -2609,6 +2609,41 @@ message GithubConnectorSpecV3 {
+    // [SAML] Attributes to roles warnings.
+    SAML_ATTRIBUTES_TO_ROLES_WARNINGS = 102;
+    // [SAML] Assertion info.
+    SAML_ASSERTION_INFO = 103;
+    // [SAML] Attribute statements.
+    SAML_ATTRIBUTE_STATEMENTS = 104;
```
What are these "types"? It does look a bit like we're trying to replace properly typed structs with this types enum + interface.

r0mant commented on lib/auth/auth_with_roles.go:
```diff
@@ -2328,18 +2329,54 @@ func (a *ServerWithRoles) CreateSAMLAuthRequest(req services.SAMLAuthRequest) (*
 
+	// require additional permissions for executing SSO test flow.
+	if req.SSOTestFlow {
+		if err := a.authConnectorAction(apidefaults.Namespace, types.KindSAML, types.VerbCreate); err != nil {
```
Why does SSO test require "create" permissions for the connector? If you're running a test, you're not actually creating the connector, right?

Tener replied:
Security perspective: we want the permissions to be high enough to argue the user is not gaining too much power by running the test.

r0mant commented on lib/auth/auth_with_roles.go:
```diff
@@ -2328,18 +2329,54 @@ func (a *ServerWithRoles) CreateSAMLAuthRequest(req services.SAMLAuthRequest) (*
+	return a.authServer.ValidateSAMLResponse(ctx, re)
+}
+
+func (a *ServerWithRoles) GetSAMLAuthRequest(ctx context.Context, id string) (*services.SAMLAuthRequest, error) {
```
Nit: Can you please add godocs on the public methods? For GetSSODiagnosticInfo below too.

r0mant commented on lib/auth/clt.go:
```diff
@@ -1751,7 +1777,13 @@ type IdentityService interface {
+	ValidateSAMLResponse(ctx context.Context, re string) (*SAMLAuthResponse, error)
+
+	// GetSAMLAuthRequest returns SAML auth request if found
+	GetSAMLAuthRequest(ctx context.Context, id string) (*services.SAMLAuthRequest, error)
```
Nit: For consistency:
```suggestion
	GetSAMLAuthRequest(ctx context.Context, authRequestID string) (*services.SAMLAuthRequest, error)
```

r0mant commented on lib/client/redirect.go:
```diff
@@ -75,10 +75,14 @@ type Redirector struct {
 	context context.Context
 	// cancel broadcasts cancel
 	cancel context.CancelFunc
+	// issueSSOLoginConsoleRequest if not nil will be used as custom implementation of IssueSSOLoginConsoleRequest function.
+	issueSSOLoginConsoleRequest IssueSSOLoginConsoleRequest
 }
 
+type IssueSSOLoginConsoleRequest func(req SSOLoginConsoleReq) (*SSOLoginConsoleResponse, error)
```
Nit: I think these types of things are usually named with "fn" or "func" suffix, I'd name it:
```suggestion
	// ssoLoginConsoleRequestFn allows to override SSO login request creation logic. Used for xxxxx.
	ssoLoginConsoleRequestFn SSOLoginConsoleRequestFunc
}

type SSOLoginConsoleRequestFunc func(req SSOLoginConsoleReq) (*SSOLoginConsoleResponse, error)
```

Tener replied:
Done: f0503dbbae9cbb592c3d347ae51b54c0b2f95235

r0mant commented on lib/client/redirect.go:
```diff
@@ -75,10 +75,14 @@ type Redirector struct {
+
 // NewRedirector returns new local web server redirector
-func NewRedirector(ctx context.Context, login SSHLoginSSO) (*Redirector, error) {
+func NewRedirector(ctx context.Context, login SSHLoginSSO, issueLogin IssueSSOLoginConsoleRequest) (*Redirector, error) {
```
Can we turn this into a RedirectorConfig struct instead of additional argument?

r0mant commented on lib/client/redirect.go:
```diff
@@ -145,32 +150,48 @@ func (rd *Redirector) Start() error {
 	return nil
 }
 
+func (rd *Redirector) IssueSSOLoginConsoleRequest(req SSOLoginConsoleReq) (*SSOLoginConsoleResponse, error) {
```
Nit: Add godoc.

Tener replied:
Done: f0503dbbae9cbb592c3d347ae51b54c0b2f95235

r0mant commented on lib/client/weblogin.go:
```diff
@@ -257,8 +257,8 @@ func initClient(proxyAddr string, insecure bool, pool *x509.CertPool) (*WebClien
 // SSHAgentSSOLogin is used by tsh to fetch user credentials using OpenID Connect (OIDC) or SAML.
-func SSHAgentSSOLogin(ctx context.Context, login SSHLoginSSO) (*auth.SSHLoginResponse, error) {
-	rd, err := NewRedirector(ctx, login)
+func SSHAgentSSOLogin(ctx context.Context, login SSHLoginSSO, issueLogin IssueSSOLoginConsoleRequest) (*auth.SSHLoginResponse, error) {
```
Like above, let's please use a config struct for this.

Tener replied:
Done: f0503dbbae9cbb592c3d347ae51b54c0b2f95235

r0mant commented on lib/services/saml.go:
```diff
@@ -95,6 +95,10 @@ func ValidateSAMLConnector(sc types.SAMLConnector) error {
 	}
 
+	if len(sc.GetAttributesToRoles()) == 0 {
+		return trace.BadParameter("attributes_to_roles is empty, authorization with connector would never give any roles.")
```
```suggestion
		return trace.BadParameter("attributes_to_roles is empty, authorization with connector would never give any roles")
```

r0mant commented on lib/auth/saml.go:
```diff
@@ -128,6 +123,41 @@ func (a *Server) CreateSAMLAuthRequest(req services.SAMLAuthRequest) (*services.
+		// stateless test flow
+		connector, err := types.NewSAMLConnector(req.ConnectorID, *req.ConnectorSpec)
+		if err != nil {
+			return nil, nil, err
```
```suggestion
			return nil, nil, trace.Wrap(err)
```

r0mant commented on lib/auth/saml.go:
```diff
@@ -128,6 +123,41 @@ func (a *Server) CreateSAMLAuthRequest(req services.SAMLAuthRequest) (*services.
+		// validate, set defaults for connector
+		err = services.ValidateSAMLConnector(connector)
+		if err != nil {
+			return nil, nil, err
```
```suggestion
			return nil, nil, trace.Wrap(err)
```

r0mant commented on lib/auth/saml.go:
```diff
@@ -148,7 +178,19 @@ func (a *Server) getSAMLProvider(conn types.SAMLConnector) (*saml2.SAMLServicePr
 }
 
-func (a *Server) calculateSAMLUser(connector types.SAMLConnector, assertionInfo saml2.AssertionInfo, request *services.SAMLAuthRequest) (*createUserParams, error) {
+func (a *Server) ssoDiagInfo(ctx context.Context, authKind string, id string, infoType types.SSOInfoType, value interface{}) {
```
Nit: `ssoDiagInfo` reads more like a variable than a function name. It's not clear: does it return ssoDiagInfo, does it set it somewhere, etc.

Tener replied:
Done.

r0mant commented on lib/auth/saml.go:
```diff
@@ -336,87 +406,131 @@ func (a *Server) ValidateSAMLResponse(samlResponse string) (*SAMLAuthResponse, e
+	user, err := a.createSAMLUser(params, request.SSOTestFlow)
 	if err != nil {
-		return re, trace.Wrap(err)
+		traceErr("Failed to create user from provided parameters.", err)
+		return nil, trace.Wrap(err)
```
Instead of doing `traceErr` in each "err != nil" branch, can you update the returned errors to include user message and add the error to the diagnostic once at the caller side?

```go
return nil, trace.Wrap(err, "failed to create user from provider parameter")
```

Tener replied:
Done: e478da1c5ac88a1dbe5f21b4af20772dc980da82

r0mant commented on lib/auth/saml.go:
```diff
@@ -158,11 +200,28 @@ func (a *Server) calculateSAMLUser(connector types.SAMLConnector, assertionInfo
+		a.ssoDiagInfo(ctx, types.KindSAML, request.ID, infoType, value)
+	}
+
+	diagInfo(types.SSOInfoType_SAML_TRAITS_FROM_ASSERTIONS, p.traits)
+	diagInfo(types.SSOInfoType_SAML_CONNECTOR_TRAIT_MAPPING, connector.GetTraitMappings())
```
These "diagInfo" calls look a bit out of place throughout the function now, I would consider refactoring/organizing it a bit like this:

- Introduce a concept of the "login context" - a struct that will be keeping the context for this login flow.
- Propagate login context to each method that needs it and will be attaching diagnostic info to it.
- Once the login is done (fail or success), save the diag info in the backend.
- I will also reiterate that I'm not a fan of using int "type" + interface, I would much prefer a bunch of typed structs even if it's a little more verbose. Or just use plain JSON.

So the flow would be similar to what happens here already, just a bit more structured IMO.

Tener replied:
Think of the `diagInfo` calls as similar to logging. This is the same pattern really:
1. We want to extract some information about stuff that is happening during execution.
2. It doesn't change the control flow in any way. We should be able to replace it with no-op and still be good.
3. It happens in potentially unrelated parts of the system.

I used explicit passing of context in the past (https://github.com/gravitational/teleport/pull/10393) and it wasn't great:
1. It felt like unnecessary complexity.
2. Passing the service boundary was a pain.
3. There were ways to get the control flow interrupted in a way that resulted in the context being simply lost and unavailable to the caller.
4. We were hurting backward compatibility by changing the APIs quite a lot.

Tener replied:
On further thought and rereading the suggestion, I think that perhaps some middle ground can be achieved. Right now all of the action is happening in `saml.go`, so maybe there is some room for refactoring here. For sure we could reduce the number of calls to the backend and drop the range iteration on retrieval, which is something.

r0mant reviewed (changes requested):
In addition to the feedback, we'd also need to add test coverage.

Tener committed:
Apply suggestions from code review

Co-authored-by: Roman Tkachenko <roman@goteleport.com>

Tener committed:
Merge branch 'master' into tener/tctl-sso

Tener committed:
Comments for exported methods.

Tener committed:
Make Redirector customizable with RedirectorConfig

Tener committed:
Better error message.

Tener committed:
Better name for function.

Tener committed:
Applying feedback from code review.

- store SSO diag info only for SSO test flow
- use `trace.AddUserMessage` for error messages
- simplify error capture: log in single place, simplified structure
- rename helper function

Tener committed:
Rename helper function.

Tener committed:
Merge branch 'master' into tener/tctl-sso

Tener committed:
Use strongly typed proto messages.

Tener committed:
Switch to structured SSO diag info, store it as single object, submit in one shot.

Tener committed:
Merge branch 'master' into tener/tctl-sso

Tener committed:
Fix import.

Tener committed:
API: `go mod tidy`

Tener committed:
Format imports.

Tener committed:
Merge branch 'master' into tener/tctl-sso

r0mant commented on api/types/types.proto:
```diff
@@ -2624,6 +2634,90 @@ message GithubConnectorSpecV3 {
 
+// SAMLAttributesToRolesWarnings contains warnings produced during the process of mapping the SAML
+// attributes to roles.
+message SAMLAttributesToRolesWarnings {
```
Does this type need to have such a specific name? Can it be named something like `SSOTestWarnings` and reused for other connector flows too?

Tener replied:
I don't know if there will be other users for such a type, but indeed we can pick a more generic name here.

r0mant commented on api/types/types.proto:
```diff
@@ -2624,6 +2634,90 @@ message GithubConnectorSpecV3 {
+    string Username = 2 [ (gogoproto.jsontag) = "username,omitempty" ];
+    // Logins is a list of available unix logins.
+    repeated string Logins = 3 [ (gogoproto.jsontag) = "logins,omitempty" ];
+    // KubeGroups is a list of assigned kube groups.
+    repeated string KubeGroups = 4 [ (gogoproto.jsontag) = "kube_groups,omitempty" ];
+    // KubeUsers is a list of available kube users.
+    repeated string KubeUsers = 5 [ (gogoproto.jsontag) = "kube_users,omitempty" ];
```
Do we need to include database users/names here also?

Tener replied:
No, those are in `Traits` field. `KubeUsers` and `KubeGroups` are something of a special (legacy) case, I expect these to go away at some point in the future.

r0mant commented on lib/auth/saml.go:
```diff
@@ -128,6 +123,41 @@ func (a *Server) CreateSAMLAuthRequest(req services.SAMLAuthRequest) (*services.
+func (a *Server) getConnectorAndProvider(ctx context.Context, req services.SAMLAuthRequest) (types.SAMLConnector, *saml2.SAMLServiceProvider, error) {
+	if req.SSOTestFlow {
+		// stateless test flow
+		connector, err := types.NewSAMLConnector(req.ConnectorID, *req.ConnectorSpec)
```
I would check that `req.ConnectorSpec` is not nil (or empty).

r0mant commented on lib/auth/saml.go:
```diff
@@ -148,21 +178,53 @@ func (a *Server) getSAMLProvider(conn types.SAMLConnector) (*saml2.SAMLServicePr
 
-func (a *Server) calculateSAMLUser(connector types.SAMLConnector, assertionInfo saml2.AssertionInfo, request *services.SAMLAuthRequest) (*createUserParams, error) {
-	var err error
+type ssoDiagContext struct {
```
Since this is probably going to be reused b/w different connector types, I think it's worth moving it into its own file.

r0mant commented on lib/auth/saml.go:
```diff
@@ -148,21 +178,53 @@ func (a *Server) getSAMLProvider(conn types.SAMLConnector) (*saml2.SAMLServicePr
+	info      types.SSODiagnosticInfo
+}
+
+func (c *ssoDiagContext) Write(ctx context.Context) {
```
Nit: Add godoc, otherwise it's not immediately clear where it "writes" it.

r0mant commented on lib/auth/saml.go:
```diff
@@ -148,21 +178,53 @@ func (a *Server) getSAMLProvider(conn types.SAMLConnector) (*saml2.SAMLServicePr
-	var err error
+type ssoDiagContext struct {
+	authKind                string
+	createSSODiagnosticInfo func(ctx context.Context, authKind string, authRequestID string, info types.SSODiagnosticInfo) error
```
This field might warrant a godoc explaining how it's used.

And come to think of it, godocs on other fields might be useful too.

r0mant commented on lib/auth/saml.go:
```diff
@@ -336,87 +415,107 @@ func (a *Server) ValidateSAMLResponse(samlResponse string) (*SAMLAuthResponse, e
+	request, err := a.Identity.GetSAMLAuthRequest(ctx, requestID)
 	if err != nil {
-		return nil, trace.Wrap(err)
+		return nil, trace.Wrap(err).AddUserMessage("Failed to get SAML Auth Request")
```
Is there a difference between `trace.Wrap(err).AddUserMessage("xxx")` and `trace.Wrap(err, "xxx")`?

Tener replied:
There is tiny, but it is in favor of using `trace.Wrap(err, "xxx")`. I'll convert the code to use that.

r0mant commented on lib/auth/saml.go:
```diff
@@ -336,87 +415,107 @@ func (a *Server) ValidateSAMLResponse(samlResponse string) (*SAMLAuthResponse, e
+	}
+
+	if diagCtx.testFlow {
+		diagCtx.info.SAMLAssertionInfo = (*types.AssertionInfoWrapper)(assertionInfo)
```
Check for nil.

r0mant commented on lib/auth/saml.go:
```diff
@@ -425,6 +524,12 @@ func (a *Server) validateSAMLResponse(samlResponse string) (*samlAuthResponse, e
 
+	// In test flow skip signing and creating web sessions.
+	if request.SSOTestFlow {
+		diagCtx.info.Success = "test flow"
```
`diagCtx.info.Success = "test flow"` looks a bit strange/random, is this "success" marker needed? If it is, why is it a string instead of a bool?

Tener replied:
I used a string for symmetry with `Error` and as a room for returning additional data. But I concur that this isn't really utilized that much and we could probably get away with just a bool. If we ever need to return extra data, we can simply add another field.

r0mant commented on lib/auth/saml.go:
```diff
@@ -434,36 +539,38 @@ func (a *Server) validateSAMLResponse(samlResponse string) (*samlAuthResponse, e
 	}
 
-	return re, nil
+	diagCtx.info.Success = "regular flow"
```
Following up on the comment above, looks like this is how you differentiate b/w "test" and "regular" flow. Maybe rather make a constant like `SSOFlowType` and use it in the diag context?

```go
type SSOFlowType string
var (
  SSORegularFlow = "regular"
  SSOTestFlow = "test"
)
```

Tener replied:
I think I'll make it simpler still: add bool `TestFlow` to `SSODiagnosticInfo` and make `Success` into `bool` too.

r0mant commented on lib/client/redirect.go:
```diff
@@ -202,6 +231,11 @@ func (rd *Redirector) callback(w http.ResponseWriter, r *http.Request) (*auth.SS
 	}
 
+	if r.URL.Query().Has("err") {
+		err := r.URL.Query().Get("err")
```
Should this check that `err` isn't empty also?

Tener replied:
If the `err` is present but empty there isn't much we can do differently here. No way to fail even harder...

r0mant commented on lib/services/local/users.go:
```diff
@@ -1215,6 +1215,62 @@ func (s *IdentityService) GetSAMLAuthRequest(id string) (*services.SAMLAuthReque
+
+	switch authKind {
+	case types.KindSAML, types.KindGithub, types.KindOIDC:
+		// nothing to do
```
```suggestion
```

Tener replied:
The comment is there to indicate this is an explicit choice to do nothing. A seasoned reader will recognize a pattern, but a comment will make it unambiguous for everyone, including less experienced readers.

r0mant commented on lib/services/local/users.go:
```diff
@@ -1215,6 +1215,62 @@ func (s *IdentityService) GetSAMLAuthRequest(id string) (*services.SAMLAuthReque
+
+	switch authKind {
+	case types.KindSAML, types.KindGithub, types.KindOIDC:
+		// nothing to do
```
```suggestion
```

Tener replied:
As in the other place: the comment is there to indicate this is an explicit choice to do nothing. A seasoned reader will recognize a pattern, but a comment will make it unambiguous for everyone, including less experienced readers.

r0mant commented on lib/services/local/users.go:
```diff
@@ -1215,6 +1215,62 @@ func (s *IdentityService) GetSAMLAuthRequest(id string) (*services.SAMLAuthReque
+// GetSSODiagnosticInfo returns SSO diagnostic info records.
+func (s *IdentityService) GetSSODiagnosticInfo(ctx context.Context, authKind string, authRequestID string) (*types.SSODiagnosticInfo, error) {
+	if authRequestID == "" {
+		return nil, trace.BadParameter("missing parameter id")
```
```suggestion
		return nil, trace.BadParameter("missing parameter authRequestID")
```

r0mant commented on lib/services/saml.go:
```diff
@@ -95,6 +95,10 @@ func ValidateSAMLConnector(sc types.SAMLConnector) error {
 	}
 
+	if len(sc.GetAttributesToRoles()) == 0 {
+		return trace.BadParameter("attributes_to_roles is empty, authorization with connector would never give any roles")
```
```suggestion
		return trace.BadParameter("attributes_to_roles is empty, authorization with connector would never assign any roles")
```

r0mant reviewed (commented):
I left a few more minor suggestions, the implementation is looking good to me now! But I think this is still missing test coverage?

Tener commented:
> I left a few more minor suggestions, the implementation is looking good to me now! But I think this is still missing test coverage?

@r0mant thanks for the review, I'll add test coverage next.

Tener commented on api/types/assertion_info_wrapper.go:
```diff
@@ -0,0 +1,54 @@
+	saml2 "github.com/russellhaering/gosaml2"
+)
+
+type AssertionInfoWrapper saml2.AssertionInfo
```
`AssertionInfo` is third party type that we want to use in .proto (`customtype` tag). This requires a couple of methods to be present, which we implement here.

I'll add the godoc explaining the same.

Tener committed:
Apply suggestions from code review

Co-authored-by: Roman Tkachenko <roman@goteleport.com>

Tener committed:
Small refinements.

- Rename auth.AssertionInfoWrapper to shorter auth.AssertionInfo.
- Add bool TestFlow to SSODiagnosticInfo
- Make SSODiagnosticInfo.Success a bool instead of string.
- Rename SAMLAttributesToRolesWarnings to more generic SSOWarnings
- Add godocs in several places.
- Avoid explicit call to trace.AddUserMessage() where possible.

Tener committed:
Add check against empty ConnectorID.

Tener committed:
Add several tests.

Tener committed:
Imports.

Tener committed:
Merge branch 'master' into tener/tctl-sso

Tener committed:
Lint.

Tener committed:
License.

Tener commented:
@r0mant I've added test coverage for the critical pieces. Let me know if you'd like me to add more coverage anywhere.

Tener committed:
Additional tests.

Tener committed:
Expand test coverage in lib/services for SAML.

Tener committed:
Merge branch 'master' into tener/tctl-sso

Tener committed:
Check all returned values in test.

r0mant reviewed (approved):
@Tener Ship it! But after 2nd approval :)

Tener committed:
Merge branch 'master' into tener/tctl-sso

Tener committed:
Cleanup go.mod post merge.

Tener committed:
Merge branch 'master' into tener/tctl-sso
